pkg/txn/storage/txn/mvcc.go

```diff
@@ -36,20 +36,27 @@ type MVCCValue[T any] struct {
 
 // Read reads the visible value from Values
 // readTime's logical time should be monotonically increasing in one transaction to reflect commands order
-func (m *MVCC[T]) Read(tx *Transaction, readTime Timestamp) *T {
+func (m *MVCC[T]) Read(tx *Transaction, readTime Timestamp) (*T, error) {
 	if tx.State.Load() != Active {
 		panic("should not call Read")
 	}
 
 	m.RLock()
 	defer m.RUnlock()
 	for i := len(m.Values) - 1; i >= 0; i-- {
-		if m.Values[i].Visible(tx.ID, readTime) {
-			return m.Values[i].Value
+		value := m.Values[i]
+		if value.Visible(tx.ID, readTime) {
+			if value.BornTx != tx && value.BornTime.Greater(tx.BeginTime) {
```

Comment:
Is this check just for extra caution?   Or it can really happen?

Reply:
It can happen. We may handle this differently depending on the isolation level we want to implement. Currently, it's serializable.  For snapshot-isolation, we may skip newly written values.